### PR TITLE
libjlm/tooling: Added missing support for -J in jlc

### PR DESCRIPTION
**Pull request by sjalander**
jlc accepted the -J argument but didn't pass on the optimizations to jlm-opt

### CODE REVIEW

File: libjlm/src/tooling/CommandGraphGenerator.cpp [hunk 1]
```diff
@@ -114,12 +114,31 @@ JlcCommandGraphGenerator::GenerateCommandGraph(const JlcCommandLineOptions & com
...
+	  {"pll", JlmOptCommand::Optimization::NodePullIn},
+	  {"red", JlmOptCommand::Optimization::NodeReduction},
+	  {"ivt", JlmOptCommand::Optimization::ThetaGammaInversion},
+	  {"url", JlmOptCommand::Optimization::LoopUnrolling},
+        });
+        for (const auto & jlmOpt : commandLineOptions.JlmOptOptimizations_)
+	{
+          JLM_ASSERT(map.find(jlmOpt) != map.end());
+          optimizations.push_back(map[jlmOpt]);
+	}
```
Comment (phate) [inline libjlm/src/tooling/CommandGraphGenerator.cpp:136]: Could we fix the identation please?

**Comment by phate** [comment]
Please add a unit test as well.

File: libjlm/src/tooling/CommandGraphGenerator.cpp (hunk 1 above)
Comment (sjalander) [reply libjlm/src/tooling/CommandGraphGenerator.cpp]: Indentation and test should be fixed

File: libjlm/src/tooling/CommandGraphGenerator.cpp [hunk 2]
```diff
@@ -114,12 +114,31 @@ JlcCommandGraphGenerator::GenerateCommandGraph(const JlcCommandLineOptions & com
...
+          {"pll", JlmOptCommand::Optimization::NodePullIn},
+          {"red", JlmOptCommand::Optimization::NodeReduction},
+          {"ivt", JlmOptCommand::Optimization::ThetaGammaInversion},
+          {"url", JlmOptCommand::Optimization::LoopUnrolling},
+        });
+        for (const auto & jlmOpt : commandLineOptions.JlmOptOptimizations_)
+	{
+          JLM_ASSERT(map.find(jlmOpt) != map.end());
+          optimizations.push_back(map[jlmOpt]);
+	}
```
Comment (phate) [inline libjlm/src/tooling/CommandGraphGenerator.cpp:136]: The parentheses are still not properly indented.

File: libjlm/include/jlm/tooling/Command.hpp [hunk 3]
```diff
@@ -384,6 +384,7 @@ class JlmOptCommand final : public Command {
 
   filepath InputFile_;
   filepath OutputFile_;
+public:
```
Comment (phate) [inline libjlm/include/jlm/tooling/Command.hpp:387]: Please provide an accessor method and keep this private:
`const std::vector<Optimization> &
Optimizations() const noexcept
{
return Optimizations_
}`

File: tests/libjlm/tooling/TestJlcCommandLineParser.cpp [hunk 4]
```diff
@@ -122,13 +123,39 @@ Test4()
...
+   * Assert
+   */
+  auto & clangCommandNode = (*commandGraph->GetEntryNode().OutgoingEdges().begin()).GetSink();
+  auto & jlmOptCommandNode = (clangCommandNode.OutgoingEdges().begin())->GetSink();
+  auto command = dynamic_cast<const jlm::JlmOptCommand*>(&jlmOptCommandNode.GetCommand());
+  assert(command->Optimizations_.size() == 2);
+  assert(command->Optimizations_[0] == jlm::JlmOptCommand::Optimization::CommonNodeElimination && \
+         command->Optimizations_[1] == jlm::JlmOptCommand::Optimization::DeadNodeElimination);
+}
+
```
Comment (phate) [inline tests/libjlm/tooling/TestJlcCommandLineParser.cpp:150]: The file is called TestJlcCommandLineParser.cpp and not TestJlcCommandGraphGenerator.cpp. Please add tests for the command graph generation in the appropriate file (or create one if none exists).

File: libjlm/src/tooling/CommandGraphGenerator.cpp (hunk 1 above)
Comment (phate) [reply libjlm/src/tooling/CommandGraphGenerator.cpp]: The parentheses of the for loop are still not properly indented.

File: tests/libjlm/tooling/TestJlcCommandLineParser.cpp (hunk 4 above)
Comment (sjalander) [reply tests/libjlm/tooling/TestJlcCommandLineParser.cpp:150]: It tests command line parsing, but to really verify it, one need to check the graph that's generated.
What name do you want for something that checks command line arguments that generate correct graphs.

Comment (phate) [reply tests/libjlm/tooling/TestJlcCommandLineParser.cpp:150]: TestJlcCommandLineParser.cpp -> Tests command line parsing. The tests take argc and argvs as input and have the CommandLineOptions as output (that is verified).

TestJlcCommandGraphGenerator -> Tests command graph generation. The tests take a CommandLineOptions as input and have the command graph as output (that is verified).

Your added test currently tests both in a file that is for command line parser tests.

File: libjlm/include/jlm/tooling/Command.hpp (hunk 3 above)
Comment (sjalander) [reply libjlm/include/jlm/tooling/Command.hpp]: Fixed

File: tests/libjlm/tooling/TestJlcCommandLineParser.cpp (hunk 4 above)
Comment (sjalander) [reply tests/libjlm/tooling/TestJlcCommandLineParser.cpp]: Fixed

File: libjlm/include/jlm/tooling/CommandLine.hpp [hunk 5]
```diff
@@ -529,4 +529,4 @@ class JhlsCommandLineParser final : public CommandLineParser {
 
 }
 
-#endif //JLM_TOOLING_COMMANDLINE_HPP
\ No newline at end of file
+#endif //JLM_TOOLING_COMMANDLINE_HPP
```
Comment (phate) [inline libjlm/include/jlm/tooling/CommandLine.hpp]: This is noise. There is no other change in this file.

File: tests/libjlm/tooling/TestJlcCommandLineParser.cpp [hunk 6]
```diff
@@ -6,6 +6,7 @@
 #include "test-registry.hpp"
 
 #include <jlm/tooling/CommandLine.hpp>
+#include <jlm/tooling/CommandGraphGenerator.hpp>
```
Comment (phate) [inline tests/libjlm/tooling/TestJlcCommandLineParser.cpp:9]: Is this include needed here? You are not doing command graph generation in this file.

File: libjlm/include/jlm/tooling/CommandLine.hpp (hunk 5 above)
Comment (sjalander) [reply libjlm/include/jlm/tooling/CommandLine.hpp]: fixed

File: tests/libjlm/tooling/TestJlcCommandLineParser.cpp (hunk 6 above)
Comment (sjalander) [reply tests/libjlm/tooling/TestJlcCommandLineParser.cpp]: fixed
